src/kbd-slide.c

```diff
@@ -0,0 +1,217 @@
+/**
+  @file kbd-slide.c
+  Code concerning state of the slide keyboard.
+
+  This file is part of ke-recv.
+
+  Copyright (C) 2020 Arthur Demchenkov <spinal.by@gmail.com>
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  version 2 as published by the Free Software Foundation.
+
+  This program is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  02110-1301 USA
+*/
+
+#include <glib.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <libevdev/libevdev.h>
+
+#include "events.h"
+#include "kbd-slide.h"
+
+/* Path to the input device directory */
+#define DEV_INPUT_PATH "/dev/input"
+/* Prefix for event files */
+#define EVENT_FILE_PREFIX "event"
+/* The name of device we are going to monitor */
+#define KBD_SLIDE_DEV_NAME "gpio_keys"
+
+static struct libevdev *kbd_slide_dev = NULL;
+static GIOChannel *kbd_slide_iochan = NULL;
+static guint kbd_slide_src_id = 0;
+
+static gboolean kbd_slide_add_handler(void);
+
+/**
+ * Try to set input device as a keypad slide handler
+ * @param path  the device file path to try
+ * @return  TRUE if device is suitable, FALSE otherwise
+ */
+static gboolean
+kbd_slide_set_input_dev(gchar *path)
+{
+    gint fd, rc;
+    struct libevdev *dev = NULL;
+
+    fd = open(path, O_RDONLY|O_NONBLOCK);
+    if (fd < 0)
+        return FALSE;
+
+    rc = libevdev_new_from_fd(fd, &dev);
+    if (rc < 0) {
+        close(fd);
+        return FALSE;
+    }
+
+    if (strcmp(libevdev_get_name(dev), KBD_SLIDE_DEV_NAME) == 0 &&
+        libevdev_has_event_type(dev, EV_SW))
+    {
+        gboolean state = libevdev_get_event_value(dev, EV_SW, SW_KEYPAD_SLIDE);
+        inform_slide_keyboard(state);
+        kbd_slide_dev = dev;
+        return TRUE;
+    }
+
+    libevdev_free(dev);
+    close(fd);
+    return FALSE;
+}
+
+/**
+ * Find and set proper input device for slide keypad monitoring
+ * @return  TRUE if found, FALSE otherwise
+ */
+static gboolean
+kbd_slide_find_evdev()
+{
+    DIR *dir = NULL;
+    struct dirent *direntry = NULL;
+    gboolean result = FALSE;
+
+    if ((dir = opendir(DEV_INPUT_PATH)) == NULL) {
+        perror(__func__);
+        return FALSE;
+    }
+
+    for (direntry = readdir(dir);
+         (direntry != NULL && telldir(dir));
+         direntry = readdir(dir))
+    {
```

Comment:
curly bracket is on a new line

Reply:
This was done on purpose. I insert newlines before curly brackets when the condition before it takes multiple lines. It helps to make the code easier to read.

Reply:
But, it differs from the coding style in the other parts of the same file. You either keep them on the same line as if() everywhere or put them on the next line, all over the place. Mixing is not a good coding style.

I agree it makes it easier to read when the bracket is on the next line - why don't you do that everywhere?

Reply:
The rules of the style I used:
```
if (...) {
  ...
  ...
}
```
```
if (... &&
    ...)
{
  ...
}
```
```
if (...)
  ...
```
I don't see a problem with the first case. It seems pretty good and easy to read.
Making like this is pretty fine too:
```
if (...)
{
  ...
  ...
}
```
But the overall code base in the package uses the first option:
```
if (...) {
  ...
  ...
}
```
So I tried to be more or less consitent with the existing code base, but placed the curly brackets after new lines in the places where it had more sense.

Reply:
The code base in the package uses this:
```
if (...
    ...) {
  ...
}
```
... and this for single commands:
```
if (...) {
  ...
}
```
I can convert the file I created to this style, if you wish.

Reply:
yes, better follow the coding style of the other files.

Reply:
I had a look at the code base one more time and found there's no style consitency at all.
There are cases with several style inconsistencies in a single file. Not to mention that
styles differ from file to file.

Some examples not to be unfounded.

camera.c:
```
            if (ioctl(fd, VIDIOC_S_FMT, &format) < 0) {
                ret = TRUE;
            }

            else {
                ret = FALSE;
            }
```
exec-func.c:
```
        if (ret != 0) {
                return FALSE;
        } else {
                return TRUE;
        }
```
swap_mgr.c:
```
      if ( !location )
         location = "";
```
swap_mgr.c:
```
   if ( size )
   {
      printf ("2. swap file available: %u MB\n", size >> 20);
   }
   else
   {
      printf ("2. no swap file available\n");
   }
```
So, I prefer to leave it as it is. The style of the file I wrote is pretty consistent.
I defined the style rules I used in the message above.

Reply:
well, I prefer to change the coding style to:
if()
{
 line 1
 line 2
}
else
  line 3

and maybe create issue "Fix other files coding style"

If you insist on keeping your "mixed mode curly braces" style, I'll leave @MerlijnWajer to decide, I already wasted too much time arguing with you on simple issues

Reply:
Just don't want to make it look uggly with something like this:
```
    if (strcmp(libevdev_get_name(dev), KBD_SLIDE_DEV_NAME) == 0 &&
        libevdev_has_event_type(dev, EV_SW)) {
        state = libevdev_get_event_value(dev, EV_SW, SW_KEYPAD_SLIDE);
        inform_slide_keyboard(state);
        kbd_slide_dev = dev;
        return TRUE;
    }
```

Reply:
I don;t like that as well, thus move that curly bracket on a new line.

Reply:
> well, I prefer to change the coding style to ...

Implemented
> and maybe create issue "Fix other files coding style"

The problem with this will be broken "git blame" functionality.

Reply:
right

Reply:
> > well, I prefer to change the coding style to ...
> 
> Implemented.

You have missed a couple of places, please fix and I will have no more objections for merging

Reply:
Do you mean this one? `do {`
Didn't find anything else.

Reply:
https://github.com/maemo-leste/ke-recv/blob/kbd-slide/src/kbd-slide.c#L137
https://github.com/maemo-leste/ke-recv/blob/kbd-slide/src/kbd-slide.c#L143
https://github.com/maemo-leste/ke-recv/blob/kbd-slide/src/kbd-slide.c#L146

Reply:
Thanks. Fixed.